# Projeto 1 - Ciência dos Dados

Nome: Felipe Blasbalg

Nome: Arthur Fuhrken

**Atenção:** Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [1]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

**Em `filename`, coloque o nome do seu arquivo de dados!**

In [2]:
import os

filename = 'Iphone.xlsx'
if filename in os.listdir():
    print(f'Encontrei o arquivo {filename}, tudo certo para prosseguir com a prova!')
else:
    print(f'Não encontrei o arquivo {filename} aqui no diretório {os.getcwd()}, será que você não baixou o arquivo?')

Encontrei o arquivo Iphone.xlsx, tudo certo para prosseguir com a prova!


Carregando a base de dados com os tweets classificados como relevantes e não relevantes:

In [3]:
train = pd.read_excel("Iphone.xlsx")
train.head(5)

Treinamento  Classificação
0  a camera do iphone 12 vai fica bem no meio do ...              1
1  eu gostei do iphone 12 agora só falta alguém p...              1
2  rt @mahguizelini: a apple tá lançando o iphone...              0
3  esquece, hoje a apple não lança o novo iphone,...              0
4  rt @orelha31: rapaziada que puder ajuda, o @in...              0

In [4]:
test = pd.read_excel(filename, sheet_name = 'Teste')
test.head(5)

Teste  Classificação
0  o ipad air novo é quadrado então certeza que o...              0
1  iphone 12 parece um fogão de 4 bocas, mas eu q...              1
2  rt @0_camila7: eu adicionando ao carrinho todo...              1
3  já quero o iphone 12, mas tem aquele negocio n...              1
4  já lançaram o iphone 12?\n\nagora da pra compr...              0

___
## Classificador automático de sentimento


Faça aqui uma descrição do seu produto e o que considerou como relevante ou não relevante na classificação dos tweets.

ESCREVA AQUI...

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [5]:
import re 

def cleanup(text):

    #import string
    punctuation = '[!-.:?;/@]' # Note que os sinais [] são delimitadores de um conjunto.
    pattern = re.compile(punctuation)
    text_subbed = re.sub(pattern, '', text)
    return text_subbed

In [6]:
train.loc[:,'Classificação'] = train['Classificação'].astype('category')
train['Classificação'].cat.categories = ["Irrelevante", "Relevante"]
train

Treinamento Classificação
0    a camera do iphone 12 vai fica bem no meio do ...     Relevante
1    eu gostei do iphone 12 agora só falta alguém p...     Relevante
2    rt @mahguizelini: a apple tá lançando o iphone...   Irrelevante
3    esquece, hoje a apple não lança o novo iphone,...   Irrelevante
4    rt @orelha31: rapaziada que puder ajuda, o @in...   Irrelevante
..                                                 ...           ...
295  rt @juju_quebrada: no twitter tu só é famosinh...   Irrelevante
296  rt @macsuellah: e eu q achava q eu ficaria bnt...   Irrelevante
297       @jbiebschocolate a apple nem anunciou iphone   Irrelevante
298  rt @costa_duda10: o iphone 12 e o fogão 4 boca...     Relevante
299  rt @matsuckjux: o anúncio do iphone 12 #applee...   Irrelevante

[300 rows x 2 columns]

In [7]:
#relevantes:
relevantes_ = train["Classificação"] == "Relevante"
relevantes = train.loc[relevantes_,["Treinamento"]].copy()
relevantes

Treinamento
0    a camera do iphone 12 vai fica bem no meio do ...
1    eu gostei do iphone 12 agora só falta alguém p...
5    mds que horrível caralho iphone não sabe mais ...
7    deviam fazer o touch id no botão de desbloquei...
13   rt @updatechartuc: o iphone12 é o cruzamento d...
..                                                 ...
285  os iphone’s estão ficando cada vez mais feios,...
289  só quero ver o iphone, o ios e saber se há fon...
291  pq diabos não colocam esse leitor de digital d...
292  ipad 8 com a12 bionic. eu não entendo por que ...
298  rt @costa_duda10: o iphone 12 e o fogão 4 boca...

[122 rows x 1 columns]

In [8]:
#irrelevantes:
irrelevantes_ = train["Classificação"] == "Irrelevante"
irrelevantes = train.loc[irrelevantes_,["Treinamento"]].copy()
irrelevantes

Treinamento
2    rt @mahguizelini: a apple tá lançando o iphone...
3    esquece, hoje a apple não lança o novo iphone,...
4    rt @orelha31: rapaziada que puder ajuda, o @in...
6    rt @jonathanwsg: brasileiro: acho que vou comp...
8    o novo ipad air essa a parecer muito o novo de...
..                                                 ...
294  meu iphone tá obsoleto desde que lançou o x se...
295  rt @juju_quebrada: no twitter tu só é famosinh...
296  rt @macsuellah: e eu q achava q eu ficaria bnt...
297       @jbiebschocolate a apple nem anunciou iphone
299  rt @matsuckjux: o anúncio do iphone 12 #applee...

[178 rows x 1 columns]

In [9]:
lista_rel = relevantes["Treinamento"].tolist()
lista_rel

['a camera do iphone 12 vai fica bem no meio do aparelho, pronto os apple maníaco vão achar isso o máximo',
 'eu gostei do iphone 12 agora só falta alguém pra me dar um de presente',
 'mds que horrível caralho iphone não sabe mais oq fazer sem o steve jobs msm https://t.co/fbu9px0hsv',
 'deviam fazer o touch id no botão de desbloqueio do iphone também',
 'rt @updatechartuc: o iphone12 é o cruzamento do iphone 5 com o iphone 11. https://t.co/equb9tmn5j',
 'off apple nunca faz nada muito inovador mas queria esse iphone no o',
 'rt @lailafajardoo: iphone 12 lança logo pra diminuir o preço dos antigos p eu comprarrrrr',
 '#appleevent só pra constar hoje é o lançamento do apple watch 6 e de um novo ipad.... iphone só em outubro',
 '@leonirpoletti olha o lançamento do iphone 12 que tu vai melhorar',
 'saudades iphone, agora eu que me contente com o meu samsung (que é bom) até ter dimdim pra comprar iphone que saco',
 'rt @necelebridades: a auto-estima do brasileiro merece elogios, pq ficar e

In [10]:
lista_irrel = irrelevantes["Treinamento"].tolist()
lista_irrel


['rt @mahguizelini: a apple tá lançando o iphone 12 e até agora não fez um fio de carregador que prestasse 🤩🤩🤩\n#appleevent https://t.co/uqkfi…',
 'esquece, hoje a apple não lança o novo iphone, e nem divulga a data do ios14 🙄',
 'rt @orelha31: rapaziada que puder ajuda, o @indiozinvive31 perdeu hoje na volta da praia dentro do brt uma mochila com dois celulares um ip…',
 'rt @jonathanwsg: brasileiro: acho que vou comprar um iphone 12\n\natendente: qual a forma de pagamento que você vai usar?\n\n#appleevent https:…',
 'o novo ipad air essa a parecer muito o novo design vazado do iphone 12 só que com um tamanho maior 👀 #appleevent',
 'rt @luisfmartinsc: o novo ipad air essa a parecer muito o novo design vazado do iphone 12 só que com um tamanho maior 👀 #appleevent',
 '@jooaobernarrdo já tão a abusar, depois do relógio costumava ser o iphone, agora tão a falar dos ipads..',
 'novo iphone está quase a ser anunciado, eu:\n#appleevent https://t.co/iqek7hba0v',
 'nee .. tem iphone é rica kkk

In [11]:
texto_rel = ''.join(lista_rel)
texto_rel

'a camera do iphone 12 vai fica bem no meio do aparelho, pronto os apple maníaco vão achar isso o máximoeu gostei do iphone 12 agora só falta alguém pra me dar um de presentemds que horrível caralho iphone não sabe mais oq fazer sem o steve jobs msm https://t.co/fbu9px0hsvdeviam fazer o touch id no botão de desbloqueio do iphone tambémrt @updatechartuc: o iphone12 é o cruzamento do iphone 5 com o iphone 11. https://t.co/equb9tmn5joff apple nunca faz nada muito inovador mas queria esse iphone no ort @lailafajardoo: iphone 12 lança logo pra diminuir o preço dos antigos p eu comprarrrrr#appleevent só pra constar hoje é o lançamento do apple watch 6 e de um novo ipad.... iphone só em outubro@leonirpoletti olha o lançamento do iphone 12 que tu vai melhorarsaudades iphone, agora eu que me contente com o meu samsung (que é bom) até ter dimdim pra comprar iphone que sacort @necelebridades: a auto-estima do brasileiro merece elogios, pq ficar empolgado com iphone 12 e #appleevent com os "precin

In [12]:
texto_irrel = ''.join(lista_irrel)
texto_irrel

'rt @mahguizelini: a apple tá lançando o iphone 12 e até agora não fez um fio de carregador que prestasse 🤩🤩🤩\n#appleevent https://t.co/uqkfi…esquece, hoje a apple não lança o novo iphone, e nem divulga a data do ios14 🙄rt @orelha31: rapaziada que puder ajuda, o @indiozinvive31 perdeu hoje na volta da praia dentro do brt uma mochila com dois celulares um ip…rt @jonathanwsg: brasileiro: acho que vou comprar um iphone 12\n\natendente: qual a forma de pagamento que você vai usar?\n\n#appleevent https:…o novo ipad air essa a parecer muito o novo design vazado do iphone 12 só que com um tamanho maior 👀 #appleeventrt @luisfmartinsc: o novo ipad air essa a parecer muito o novo design vazado do iphone 12 só que com um tamanho maior 👀 #appleevent@jooaobernarrdo já tão a abusar, depois do relógio costumava ser o iphone, agora tão a falar dos ipads..novo iphone está quase a ser anunciado, eu:\n#appleevent https://t.co/iqek7hba0vnee .. tem iphone é rica kkkkkkkk https://t.co/ceugciesnuo ipad quadr

In [13]:
rel_clean = cleanup(texto_rel)
rel_clean

'a camera do iphone 12 vai fica bem no meio do aparelho pronto os apple maníaco vão achar isso o máximoeu gostei do iphone 12 agora só falta alguém pra me dar um de presentemds que horrível caralho iphone não sabe mais oq fazer sem o steve jobs msm httpstcofbu9px0hsvdeviam fazer o touch id no botão de desbloqueio do iphone tambémrt updatechartuc o iphone12 é o cruzamento do iphone 5 com o iphone 11 httpstcoequb9tmn5joff apple nunca faz nada muito inovador mas queria esse iphone no ort lailafajardoo iphone 12 lança logo pra diminuir o preço dos antigos p eu comprarrrrrappleevent só pra constar hoje é o lançamento do apple watch 6 e de um novo ipad iphone só em outubroleonirpoletti olha o lançamento do iphone 12 que tu vai melhorarsaudades iphone agora eu que me contente com o meu samsung que é bom até ter dimdim pra comprar iphone que sacort necelebridades a autoestima do brasileiro merece elogios pq ficar empolgado com iphone 12 e appleevent com os precinhos camaradas…bakachristino fal

In [14]:
irrel_clean = cleanup(texto_irrel)
irrel_clean

'rt mahguizelini a apple tá lançando o iphone 12 e até agora não fez um fio de carregador que prestasse 🤩🤩🤩\nappleevent httpstcouqkfi…esquece hoje a apple não lança o novo iphone e nem divulga a data do ios14 🙄rt orelha31 rapaziada que puder ajuda o indiozinvive31 perdeu hoje na volta da praia dentro do brt uma mochila com dois celulares um ip…rt jonathanwsg brasileiro acho que vou comprar um iphone 12\n\natendente qual a forma de pagamento que você vai usar\n\nappleevent https…o novo ipad air essa a parecer muito o novo design vazado do iphone 12 só que com um tamanho maior 👀 appleeventrt luisfmartinsc o novo ipad air essa a parecer muito o novo design vazado do iphone 12 só que com um tamanho maior 👀 appleeventjooaobernarrdo já tão a abusar depois do relógio costumava ser o iphone agora tão a falar dos ipadsnovo iphone está quase a ser anunciado eu\nappleevent httpstcoiqek7hba0vnee  tem iphone é rica kkkkkkkk httpstcoceugciesnuo ipad quadraro putz o iphone vazado é real appleevento f

In [15]:
texto_total = rel_clean + irrel_clean
texto_total

'a camera do iphone 12 vai fica bem no meio do aparelho pronto os apple maníaco vão achar isso o máximoeu gostei do iphone 12 agora só falta alguém pra me dar um de presentemds que horrível caralho iphone não sabe mais oq fazer sem o steve jobs msm httpstcofbu9px0hsvdeviam fazer o touch id no botão de desbloqueio do iphone tambémrt updatechartuc o iphone12 é o cruzamento do iphone 5 com o iphone 11 httpstcoequb9tmn5joff apple nunca faz nada muito inovador mas queria esse iphone no ort lailafajardoo iphone 12 lança logo pra diminuir o preço dos antigos p eu comprarrrrrappleevent só pra constar hoje é o lançamento do apple watch 6 e de um novo ipad iphone só em outubroleonirpoletti olha o lançamento do iphone 12 que tu vai melhorarsaudades iphone agora eu que me contente com o meu samsung que é bom até ter dimdim pra comprar iphone que sacort necelebridades a autoestima do brasileiro merece elogios pq ficar empolgado com iphone 12 e appleevent com os precinhos camaradas…bakachristino fal

In [16]:
texto_rel_s = rel_clean.split()
texto_rel_s

['a',
 'camera',
 'do',
 'iphone',
 '12',
 'vai',
 'fica',
 'bem',
 'no',
 'meio',
 'do',
 'aparelho',
 'pronto',
 'os',
 'apple',
 'maníaco',
 'vão',
 'achar',
 'isso',
 'o',
 'máximoeu',
 'gostei',
 'do',
 'iphone',
 '12',
 'agora',
 'só',
 'falta',
 'alguém',
 'pra',
 'me',
 'dar',
 'um',
 'de',
 'presentemds',
 'que',
 'horrível',
 'caralho',
 'iphone',
 'não',
 'sabe',
 'mais',
 'oq',
 'fazer',
 'sem',
 'o',
 'steve',
 'jobs',
 'msm',
 'httpstcofbu9px0hsvdeviam',
 'fazer',
 'o',
 'touch',
 'id',
 'no',
 'botão',
 'de',
 'desbloqueio',
 'do',
 'iphone',
 'tambémrt',
 'updatechartuc',
 'o',
 'iphone12',
 'é',
 'o',
 'cruzamento',
 'do',
 'iphone',
 '5',
 'com',
 'o',
 'iphone',
 '11',
 'httpstcoequb9tmn5joff',
 'apple',
 'nunca',
 'faz',
 'nada',
 'muito',
 'inovador',
 'mas',
 'queria',
 'esse',
 'iphone',
 'no',
 'ort',
 'lailafajardoo',
 'iphone',
 '12',
 'lança',
 'logo',
 'pra',
 'diminuir',
 'o',
 'preço',
 'dos',
 'antigos',
 'p',
 'eu',
 'comprarrrrrappleevent',
 'só',
 'pra

In [17]:
texto_irrel_s = irrel_clean.split()
texto_irrel_s

['rt',
 'mahguizelini',
 'a',
 'apple',
 'tá',
 'lançando',
 'o',
 'iphone',
 '12',
 'e',
 'até',
 'agora',
 'não',
 'fez',
 'um',
 'fio',
 'de',
 'carregador',
 'que',
 'prestasse',
 '🤩🤩🤩',
 'appleevent',
 'httpstcouqkfi…esquece',
 'hoje',
 'a',
 'apple',
 'não',
 'lança',
 'o',
 'novo',
 'iphone',
 'e',
 'nem',
 'divulga',
 'a',
 'data',
 'do',
 'ios14',
 '🙄rt',
 'orelha31',
 'rapaziada',
 'que',
 'puder',
 'ajuda',
 'o',
 'indiozinvive31',
 'perdeu',
 'hoje',
 'na',
 'volta',
 'da',
 'praia',
 'dentro',
 'do',
 'brt',
 'uma',
 'mochila',
 'com',
 'dois',
 'celulares',
 'um',
 'ip…rt',
 'jonathanwsg',
 'brasileiro',
 'acho',
 'que',
 'vou',
 'comprar',
 'um',
 'iphone',
 '12',
 'atendente',
 'qual',
 'a',
 'forma',
 'de',
 'pagamento',
 'que',
 'você',
 'vai',
 'usar',
 'appleevent',
 'https…o',
 'novo',
 'ipad',
 'air',
 'essa',
 'a',
 'parecer',
 'muito',
 'o',
 'novo',
 'design',
 'vazado',
 'do',
 'iphone',
 '12',
 'só',
 'que',
 'com',
 'um',
 'tamanho',
 'maior',
 '👀',
 'applee

In [18]:
texto_total_s = texto_total.split()
texto_total_s

['a',
 'camera',
 'do',
 'iphone',
 '12',
 'vai',
 'fica',
 'bem',
 'no',
 'meio',
 'do',
 'aparelho',
 'pronto',
 'os',
 'apple',
 'maníaco',
 'vão',
 'achar',
 'isso',
 'o',
 'máximoeu',
 'gostei',
 'do',
 'iphone',
 '12',
 'agora',
 'só',
 'falta',
 'alguém',
 'pra',
 'me',
 'dar',
 'um',
 'de',
 'presentemds',
 'que',
 'horrível',
 'caralho',
 'iphone',
 'não',
 'sabe',
 'mais',
 'oq',
 'fazer',
 'sem',
 'o',
 'steve',
 'jobs',
 'msm',
 'httpstcofbu9px0hsvdeviam',
 'fazer',
 'o',
 'touch',
 'id',
 'no',
 'botão',
 'de',
 'desbloqueio',
 'do',
 'iphone',
 'tambémrt',
 'updatechartuc',
 'o',
 'iphone12',
 'é',
 'o',
 'cruzamento',
 'do',
 'iphone',
 '5',
 'com',
 'o',
 'iphone',
 '11',
 'httpstcoequb9tmn5joff',
 'apple',
 'nunca',
 'faz',
 'nada',
 'muito',
 'inovador',
 'mas',
 'queria',
 'esse',
 'iphone',
 'no',
 'ort',
 'lailafajardoo',
 'iphone',
 '12',
 'lança',
 'logo',
 'pra',
 'diminuir',
 'o',
 'preço',
 'dos',
 'antigos',
 'p',
 'eu',
 'comprarrrrrappleevent',
 'só',
 'pra

In [19]:
dataframe_rel = pd.Series(texto_rel_s)
dataframe_rel

0                        a
1                   camera
2                       do
3                   iphone
4                       12
               ...        
1848                  caro
1849                    do
1850                 mundo
1851            appleevent
1852    httpstcov4e3oqcdmp
Length: 1853, dtype: object

In [20]:
dataframe_irrel = pd.Series(texto_irrel_s)
dataframe_irrel

0                 rt
1       mahguizelini
2                  a
3              apple
4                 tá
            ...     
2675               e
2676               a
2677          vacina
2678              da
2679           covid
Length: 2680, dtype: object

In [33]:
dataframe_total = pd.Series(texto_total_s)
dataframe_total

0            a
1       camera
2           do
3       iphone
4           12
         ...  
4527         e
4528         a
4529    vacina
4530        da
4531     covid
Length: 4532, dtype: object

In [22]:
freq_abs_rel = dataframe_rel.value_counts().round(4)
freq_abs_rel

iphone        124
o              77
que            45
do             45
de             44
             ... 
comentados      1
precinhos       1
aínda           1
chorar          1
todos           1
Length: 712, dtype: int64

In [23]:
freq_abs_irrel = dataframe_irrel.value_counts().round(4)
freq_abs_irrel

iphone             166
o                  122
que                 76
e                   57
do                  55
                  ... 
tirarimjuliyear      1
itspedrito           1
🤡rt                  1
valendose            1
tesla                1
Length: 1013, dtype: int64

In [34]:
freq_abs_total = dataframe_total.value_counts().round(4)
freq_abs_total

iphone       290
o            199
que          121
do           100
de            95
            ... 
jogadorrt      1
max            1
deseja         1
crescendo      1
fazendo        1
Length: 1455, dtype: int64

In [29]:
#Função Naive Bayes
lista_prob_rel = []
for i in test['Teste']:
    P_R = 1
    i = i.split()
    for j in i:
        if j in freq_abs_rel:
            c = freq_abs_rel[j]        
        else:
            c = 0
        probabilidade = (c + 1)/(len(dataframe_rel) + len(freq_abs_total))
        P_R *= probabilidade
    lista_prob_rel.append(P_R)
lista_prob_rel

[4.4912902149084354e-34,
 1.964765921761525e-29,
 1.865251290719864e-64,
 5.907433238461953e-39,
 3.8855210868755913e-28,
 2.3170163812353594e-23,
 7.54049578889596e-39,
 1.1583101565029524e-51,
 5.949180162596768e-28,
 1.6362366804915739e-15,
 1.7620007121802136e-58,
 5.0821863180469843e-51,
 6.129032111129804e-37,
 1.1094421202642258e-37,
 1.6805222886530264e-15,
 9.638420368604777e-20,
 4.201149448154755e-27,
 8.344993348327912e-32,
 2.729807061583206e-72,
 7.338695384264874e-41,
 3.3944002167695083e-18,
 2.8848285000080893e-68,
 1.939452644877175e-58,
 1.0495154057310455e-45,
 1.964392432908988e-34,
 3.5411085699372974e-47,
 2.826204053608387e-20,
 4.1497790145202395e-31,
 2.1730634919508075e-21,
 3.166951804792072e-54,
 1.7347422165594583e-12,
 7.642306185892854e-64,
 5.293914341972314e-48,
 8.310275053778702e-16,
 2.4345370218518728e-30,
 9.14793537248457e-49,
 2.152749928287002e-46,
 7.937363988311532e-42,
 7.109355880917046e-24,
 3.964238659685266e-23,
 4.411958110519935e-61,
 

In [31]:
lista_prob_irrel = []
for i in test['Teste']:
    P_IR = 1
    i = i.split()
    for j in i:
        if j in freq_abs_irrel:
            c = freq_abs_irrel[j]        
        else:
            c = 0
        prob = (c + 1)/(len(dataframe_irrel) + len(freq_abs_total))
        P_IR *= prob
    lista_prob_irrel.append(P_IR)
lista_prob_irrel

[3.74051066319552e-36,
 9.849520665599016e-33,
 1.2877356876665404e-66,
 1.0426856657476235e-39,
 1.8994582593947565e-29,
 5.523652246803727e-23,
 5.5650816735638035e-39,
 6.656221798139138e-55,
 3.269845390156458e-29,
 2.2332208212004664e-16,
 1.0840405547884766e-60,
 2.1841364177996793e-53,
 6.17188483129455e-36,
 7.492154668228027e-39,
 6.084618924429317e-16,
 2.6547022010067316e-19,
 3.00001543083454e-29,
 1.701958961077785e-32,
 6.794389976216053e-75,
 2.910225769895044e-42,
 1.5957764622007064e-18,
 5.347743053100681e-69,
 1.267378988829918e-60,
 6.222129974133804e-47,
 6.103588411721286e-37,
 5.344206073560446e-47,
 6.556958694870568e-21,
 1.1486343032018282e-30,
 1.4332521429168507e-21,
 1.874968973391515e-55,
 8.102176025171799e-13,
 1.2461265692861447e-64,
 1.8769894502261088e-48,
 8.63449248699504e-16,
 4.878545901016722e-31,
 3.6179782933568114e-49,
 7.583321642600487e-50,
 2.9720742580373088e-43,
 1.64665917154969e-24,
 1.944626069258682e-23,
 6.994734058192715e-63,
 1.064

In [52]:
Naive_Bayes = []
for i in range(0,len(lista_prob_rel)):
    if (lista_prob_rel[i] > lista_prob_irrel[i]):
        Naive_Bayes.append('Relevante') 
    else:
        Naive_Bayes.append('Irrelevante') 
        
test["Naive Bayes"]= Naive_Bayes
test.head(40)

Teste  Classificação  \
0   o ipad air novo é quadrado então certeza que o...              0   
1   iphone 12 parece um fogão de 4 bocas, mas eu q...              1   
2   rt @0_camila7: eu adicionando ao carrinho todo...              1   
3   já quero o iphone 12, mas tem aquele negocio n...              1   
4   já lançaram o iphone 12?\n\nagora da pra compr...              0   
5      eu com o iphone xr e a apple lançando o 12🤡🤡🤡🤡              0   
6   daqui uns dias vai lançar iphone 12 , num tive...              0   
7   rt @leolaff_: esperando lançar o iphone 12 pra...              0   
8   @darkmotherchaos iphone é bom a única coisa ru...              1   
9            o iphone desse ano é o se2!!!#appleevent              0   
10  touch id no power button, interessante... será...              1   
11  rt @marvelcruel: @updatechartuc amei uc, pena ...              1   
12  consegui nem sair do iphone 8 plus ainda é lá ...              0   
13  rt @danilokauaan: algum aluno porto pra compra...              1   
14        @schdoug nem apresentaram o iphone ainda...              0   
15          #appleevent será que o iphone vem hoje ??              0   
16   faz um seguinte\n\nme da um iphone q ta td certo              1   
17  na espera do lançamento do iphone 30 p eu cons...              1   
18  sempre falei mal de iphone pq nunca tive condi...              1   
19  rt @taisrodrigues08: esperando lançar o iphone...              1   
20        @raqaudi vai lançar junto com o novo iphone              0   
21  rt @portalescutai: ainda na seção de ipads, a ...              0   
22  jessica comprou o iphone 11 e ta jogando na mi...              0   
23  rt @gabirelllys: tenho muita curiosidade em ip...              1   
24  @vitormrtns ficou lindo este grande iphone 5 😅...              1   
25  @claudifih nunca tive\nso macbook e iphone\nam...              0   
26  tem algum app assim pra iphone? https://t.co/6...              0   
27  meu iphone se quando for atualizado para o ios...              0   
28  lança logo a porra do iphone 12 q sacooooooooo...              1   
29  rt @mmvelozo_: eu indo ver o preço do iphone 1...              1   
30                   tentando entender esse iphone 12              0   
31  la vem outro iphone e eu aqui com o meu pobre ...              1   
32  @peedrohenry_ alá seu iphone copiando umas coi...              0   
33              vai ter iphone ou não???? #appleevent              0   
34  meu deus a apple ta arrebentando, não me decep...              1   
35  iphone 12 vai ser lançado e eu lembro q ainda ...              1   
36  eu aguardando o lançamento do ios 14 enquanto ...              1   
37  @comoumaddeusa só os anúncios nos sites '' tro...              1   
38  ja sabemos as cores do novo iphone rsrsrs http...              0   
39  hmmm então vai ter no iphone 12 também https:/...              0   

    Naive Bayes  
0     Relevante  
1     Relevante  
2     Relevante  
3     Relevante  
4     Relevante  
5   Irrelevante  
6     Relevante  
7     Relevante  
8     Relevante  
9     Relevante  
10    Relevante  
11    Relevante  
12  Irrelevante  
13    Relevante  
14    Relevante  
15  Irrelevante  
16    Relevante  
17    Relevante  
18    Relevante  
19    Relevante  
20    Relevante  
21    Relevante  
22    Relevante  
23    Relevante  
24    Relevante  
25  Irrelevante  
26    Relevante  
27  Irrelevante  
28    Relevante  
29    Relevante  
30    Relevante  
31    Relevante  
32    Relevante  
33  Irrelevante  
34    Relevante  
35    Relevante  
36    Relevante  
37    Relevante  
38    Relevante  
39    Relevante

In [47]:
tabela = pd.crosstab(test["Classificação"], test["Naive Bayes"], normalize = "columns")*100
tabela

Naive Bayes    Irrelevante  Relevante
Classificação                        
0                45.714286  38.604651
1                54.285714  61.395349

Falsos positivos = 38.6%

Falsos negativo = 54.3%

Verdadeiros positivos = 61.4%

Verdadeiros Negativos = 45.7%

___
### Concluindo

Analisando os resultados finais o projeto, é possível concluir que o objetivo principal do mesmo não foi atingido. Tal conclusão é baseada no fato de que os resultados finais obtidos em relação aos "Verdadeiros falsos" e "Verdadeiros negativos" não estão no padrão desejado. Isso porque uma porcentagem de acerto nas relevâncias dos tweets foi de apenas 61.4%, o que é considerada pelos membros do grupo como uma porcentagem baixa para a validação do projeto. Tal resultado pode ter sido atingido, além de inúmeras variáveis possíveis, pelo nível estabelecido na classificação manual dos tweets, sendo algo que, com o resultado final em mãos, poderia ter sido aprimorado em busca de uma maior precisão no modelo.

Além disso, algumas mensagens que contém opiniões bipolares, sarcasmo, ironia, entre outros, não foram classificados de maneira correta, sendo um possível foco de uma segunda iteração. Como maneira de solucionar este problema, sugerimos um novo modelo que considera o contexto, possíveis opiniões bipolares e/ou irônicas de modo que aborde os tweets em questão da maneira adequada, os classificando de forma que os mesmos sejam interpretados corretamente. De tal maneira, após uma segunda análise do modelo, seria possível obter uma melhor acurácia do modelo, tendo em vista que muitos dos tweets classificados estão equivodados e em locais inapropriádos levando em consideração suas posições verdadeiras.

Tendo tais pontos em vista, é de opinião mútua entre os integrantes do grupo, que nosso projeto deveria ser sim financiado, de forma que assim seria possível uma melhor análise e classificação dos tweets para uma conclusão mais acurada, sendo possível, com um budget maior, construir modelos mais sofisticados que classifiquem as mensagens de forma correta. Assim, seria possível, considerando um budget maior com mais horas de trabalho, criar mais categorias com análises mais profundas do conteúdo analisado para obter a maior acurácia possível e então, ter melhores resultados.

Assim, o classificador Naive-Bayes poderia ser utilizado para vários outros casos, uma vez que, dentro de suas proporções, fornece uma boa análise dos dados disponibilizados. Concluímos que bons exemplos de possíveis usos são de recomendar produtos para um consumidor, uma vez que tais análises são básicas e não necessitam de modelos mais profundos. Tal exemplo de classificador recomendaria produtos ao usuário com base em compras anteriores, interesses, características e dados pessoais de forma a recomendar o melhor produto possível ao usuário e ser efetivo na sugestão.

___
## Aperfeiçoamento:

Os trabalhos vão evoluir em conceito dependendo da quantidade de itens avançados:

* Limpar: \n, :, ", ', (, ), etc SEM remover emojis
* Corrigir separação de espaços entre palavras e emojis ou entre emojis e emojis
* Propor outras limpezas e transformações que não afetem a qualidade da informação ou classificação
* Criar categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante (3 categorias: C, mais categorias conta para B)
* Explicar por que não posso usar o próprio classificador para gerar mais amostras de treinamento
* Propor diferentes cenários para Naïve Bayes fora do contexto do projeto
* Sugerir e explicar melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)
* Montar um dashboard que realiza análise de sentimento e visualiza estes dados

___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**